playwright/src/test/java/com/microsoft/playwright/TestRouteWebSocket.java

```diff
@@ -0,0 +1,180 @@
+    page.evaluate("() => window.ws.send('hi')");
+    assertEquals("hi", messagePromise.get());
+    ws.close(1008, "oops");
+    page.waitForCondition(() -> {
+      Integer result = (Integer) page.evaluate("window.ws.readyState");
+      return result == 3;
+    }, new Page.WaitForConditionOptions().setTimeout(10_000));
+
+    assertEquals(
+      asList("open", "message: data=hello origin=ws://localhost:" + webSocketServer.getPort() + " lastEventId=",
+          "close code=1008 reason=oops wasClean=true"),
+      page.evaluate("window.log"));
+  }
+
+
+  @ParameterizedTest
+  @ValueSource(strings = {"no-mock", "no-match", "pass-through"})
+  public void shouldWorkWithBinaryTypeBlob(String mock, Page page) throws Exception {
+    setupRoute(page, mock);
+    Future<WebSocket> wsPromise = webSocketServer.waitForWebSocket();
+    setupWS(page, webSocketServer.getPort(), "blob");
+    org.java_websocket.WebSocket ws = wsPromise.get();
+    ws.send("hi".getBytes(StandardCharsets.UTF_8));
+    page.waitForCondition(() -> {
+      Boolean result = (Boolean) page.evaluate("() => window.log.length >= 2");
+      return result;
+    }, new Page.WaitForConditionOptions().setTimeout(10_000));
+
+    assertEquals(
+        asList("open", "message: data=blob:hi origin=ws://localhost:" + webSocketServer.getPort() + " lastEventId="),
+        page.evaluate("window.log"));
+    Future<String> messagePromise = webSocketServer.waitForMessage();
+    page.evaluate("() => window.ws.send(new Blob([new Uint8Array(['h'.charCodeAt(0), 'i'.charCodeAt(0)])]))");
+    // Without this the blob message is not sent in pass-through!
+    assertEquals(1, page.evaluate("window.ws.readyState"));
+    assertEquals("hi", messagePromise.get());
+  }
+
+  @ParameterizedTest
+  @ValueSource(strings = {"no-mock", "no-match", "pass-through"})
+  public void shouldWorkWithBinaryTypeArrayBuffer(String mock, Page page) throws Exception {
+    setupRoute(page, mock);
+    Future<WebSocket> wsPromise = webSocketServer.waitForWebSocket();
+    setupWS(page, webSocketServer.getPort(), "arraybuffer");
+    org.java_websocket.WebSocket ws = wsPromise.get();
+    ws.send("hi".getBytes(StandardCharsets.UTF_8));
+    page.waitForCondition(() -> {
+      Boolean result = (Boolean) page.evaluate("() => window.log.length >= 2");
+      return result;
+    }, new Page.WaitForConditionOptions().setTimeout(10_000));
+
+    assertEquals(
+      asList("open", "message: data=arraybuffer:hi origin=ws://localhost:" + webSocketServer.getPort() + " lastEventId="),
+      page.evaluate("window.log"));
+    Future<String> messagePromise = webSocketServer.waitForMessage();
+    page.evaluate("() => window.ws.send(new Blob([new Uint8Array(['h'.charCodeAt(0), 'i'.charCodeAt(0)])]))");
+    // Without this the blob message is not sent in pass-through!
+    assertEquals(1, page.evaluate("window.ws.readyState"));
+    assertEquals("hi", messagePromise.get());
+  }
+}
```

Comment:
I think it would be valuable to also port "should work with server" and/or "should work without server", because these tests actually exercise all of the client APIs.

Reply:
added